oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -223,12 +220,12 @@ impl Runtime {
             return Ok(());
         }
 
-        let nodes = self.nodes.read().unwrap();
+        let node_infos = self.node_infos.read().unwrap();
```

Comment:
(pre-existing) This lock scope extends to the end of the method, which is needed because of the use of the `NodeInfo` that's owned by the `node_infos` field.  I wonder if a `with_node_info(fn)` accessor might make this pattern clearer/safer?

Reply:
I was thinking about something analogous to the `ChannelMapping::with_channel` method; the function would just be the code in the rest of the block, but with the scope of lock & reference ownership clearer:
```Rust

    fn with_node_info<U, F: FnOnce(&NodeInfo) -> U>(&self, node_id: NodeId, f: F) -> U {
        let node_infos = self.node_infos.read().unwrap();
        let node_info = node_infos
            .get(&node_id)
            .expect("Invalid node_id passed into with_node_info");
        f(node_info)
    }

    /// Validate the [`NodeId`] has access to [`Handle`], returning `Err(OakStatus::ErrBadHandle)`
    /// if access is not allowed.
    fn validate_handle_access(&self, node_id: NodeId, handle: Handle) -> Result<(), OakStatus> {
        // Allow RUNTIME_NODE_ID access to all handles.
        if node_id == RUNTIME_NODE_ID {
            return Ok(());
        }

        self.with_node_info(node_id, |node_info| {
            let tracked_handles = node_info.handles.lock().unwrap();

            // Check the handle exists in the handles associated with a node, otherwise
            // return ErrBadHandle.
            if tracked_handles.contains(&handle) {
                Ok(())
            } else {
                error!(
                    "validate_handle_access: handle {:?} not found in node {:?}",
                    handle, node_id
                );
                Err(OakStatus::ErrBadHandle)
            }
        })
    }
```

Probably for another day though (maybe under #779)

Reply:
This seems like it is just re-implementing lexical scoping? I agree with the sentiment, but I think we should instead be more conscious of how we scope variables (especially lock guards), possibly by adding explicit anonymous blocks to delimit scopes (with appropriate comments). But happy to keep discussing separately.

Reply:
@daviddrysdale what do you think about going even further and turning even e.g. `let tracked_handles = node_info.handles.lock().unwrap();` into closure passing style (e.g. `node_info.with_handles(|handles|{handles.containts(handle)})`). Some consistency around locking patterns might be nice, but I'm not sure if its worth going this far? 

I'm also interested if scoping locks under lifetimes could be more ergonomic, something like:

~~~rust 
    pub fn with_all_channels<'a>(
        &'a self,
    ) -> std::sync::RwLockReadGuard<'a, HashMap<ChannelId, Channel>> {
        self.channels.read().unwrap()
    }

    // this would need something like MappedRwLockReadGuard from lock_api: https://docs.rs/lock_api/0.3.3/lock_api/struct.MappedRwLockReadGuard.html
    pub fn with_channel<'a>(
        &'a self,
        channel_id: ChannelId,
    ) -> MappedRwLockReadGuard<'a, Result<Channel, OakStatus>> {
        self.channels.read().unwrap().map(|channels|{
            channels.get(&channel_id).ok_or(OakStatus::ErrBadHandle)
        }
    }
~~~